### PR TITLE
torch.compile DTensor E2E

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR updates DTensor to support torch.compile

Cool stuff: there are some new tests in `test_dtensor.py` that show both the forward and backward graphs that we can send to inductor, when running a matmul with DTensor's. In particular, for this user code:
```
        def fn(x, y):
            dt = DTensor.from_local(x.reshape(2, 4), mesh, [Shard(0)], run_check=False)
            dt2 = DTensor.from_local(y.reshape(4, 2), mesh, [Shard(1)], run_check=False)
            dt_out = torch.matmul(dt, dt2)
            dt_out_redistribute = dt_out.redistribute(mesh, [Replicate()])
            return dt_out.to_local()
```

We generate the following fw and backward graphs.

Forward graph:
```
def forward(self, primals_1, primals_2):
    view = torch.ops.aten.view.default(primals_1, [2, 4]);  primals_1 = None
    _to_copy = torch.ops.aten._to_copy.default(view, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  view = None
    detach = torch.ops.aten.detach.default(_to_copy);  _to_copy = None
    detach_1 = torch.ops.aten.detach.default(detach);  detach = None
    view_1 = torch.ops.aten.view.default(primals_2, [4, 2]);  primals_2 = None
    _to_copy_1 = torch.ops.aten._to_copy.default(view_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  view_1 = None
    detach_2 = torch.ops.aten.detach.default(_to_copy_1);  _to_copy_1 = None
    detach_3 = torch.ops.aten.detach.default(detach_2);  detach_2 = None
    detach_4 = torch.ops.aten.detach.default(detach_1)
    all_gather_into_tensor = torch.ops.c10d_functional.all_gather_into_tensor.default(detach_3, 'ptd:0', [0, 1], 2)
    wait_tensor = torch.ops.c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    split = torch.ops.aten.split.Tensor(wait_tensor, 4);  wait_tensor = None
    getitem = split[0]
    getitem_1 = split[1];  split = None
    cat = torch.ops.aten.cat.default([getitem, getitem_1], 1);  getitem = getitem_1 = None
    detach_5 = torch.ops.aten.detach.default(cat);  cat = None
    mm = torch.ops.aten.mm.default(detach_4, detach_5);  detach_4 = detach_5 = None
    detach_6 = torch.ops.aten.detach.default(mm);  mm = None
    detach_9 = torch.ops.aten.detach.default(detach_6);  detach_6 = None
    detach_10 = torch.ops.aten.detach.default(detach_9);  detach_9 = None
    t = torch.ops.aten.t.default(detach_1);  detach_1 = None
    detach_13 = torch.ops.aten.detach.default(t);  t = None
    t_1 = torch.ops.aten.t.default(detach_3);  detach_3 = None
    detach_15 = torch.ops.aten.detach.default(t_1);  t_1 = None
    clone = torch.ops.aten.clone.default(detach_15, memory_format = torch.contiguous_format);  detach_15 = None
    return [detach_10, detach_13, clone]
```

Backward graph:
```
def forward(self, detach_13, clone, tangents_1):
    detach_11 = torch.ops.aten.detach.default(tangents_1);  tangents_1 = None
    detach_12 = torch.ops.aten.detach.default(detach_11);  detach_11 = None
    mm_1 = torch.ops.aten.mm.default(detach_13, detach_12);  detach_13 = None
    detach_14 = torch.ops.aten.detach.default(mm_1);  mm_1 = None
    detach_16 = torch.ops.aten.detach.default(detach_12);  detach_12 = None
    all_gather_into_tensor_2 = torch.ops.c10d_functional.all_gather_into_tensor.default(clone, 'ptd:0', [0, 1], 2);  clone = None
    wait_tensor_2 = torch.ops.c10d_functional.wait_tensor.default(all_gather_into_tensor_2);
    detach_17 = torch.ops.aten.detach.default(wait_tensor_2);  wait_tensor_2 = None
    mm_2 = torch.ops.aten.mm.default(detach_16, detach_17);  detach_16 = detach_17 = None
    detach_18 = torch.ops.aten.detach.default(mm_2);  mm_2 = None
    split_1 = torch.ops.aten.split.Tensor(detach_14, 2, 1);  detach_14 = None
    getitem_2 = split_1[0]
    getitem_3 = split_1[1];  split_1 = None
    cat_1 = torch.ops.aten.cat.default([getitem_2, getitem_3]);  getitem_2 = getitem_3 = None
    reduce_scatter_tensor = torch.ops.c10d_functional.reduce_scatter_tensor.default(cat_1, 'SUM', 'ptd:0', [0, 1], 2);  cat_1 = None
    wait_tensor_3 = torch.ops.c10d_functional.wait_tensor.default(reduce_scatter_tensor);  reduce_scatter_tensor = None
    detach_19 = torch.ops.aten.detach.default(wait_tensor_3);  wait_tensor_3 = None
    detach_20 = torch.ops.aten.detach.default(detach_19);  detach_19 = None
    detach_21 = torch.ops.aten.detach.default(detach_20);  detach_20 = None
    detach_22 = torch.ops.aten.detach.default(detach_21);  detach_21 = None
    _to_copy_2 = torch.ops.aten._to_copy.default(detach_22, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'));  detach_22 = None
    view_2 = torch.ops.aten.view.default(_to_copy_2, [8]);  _to_copy_2 = None
    detach_23 = torch.ops.aten.detach.default(detach_18);  detach_18 = None
    detach_24 = torch.ops.aten.detach.default(detach_23);  detach_23 = None
    _to_copy_3 = torch.ops.aten._to_copy.default(detach_24, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'));  detach_24 = None
    view_3 = torch.ops.aten.view.default(_to_copy_3, [8]);  _to_copy_3 = None
    return [view_3, view_2]
```

Some of the stuff in this graph looks kinda of silly though (e.g. an unnecessary split() + cat(), and all the extra detach() calls).

Stuff that's broken:
- functionalization is pretty horribly broken. In particular, the original strategy I used in this stack was to have functionalization run **above** subclass desugaring. But that doesn't play well with with the way we want to compile DTensor. DTensor has a few API's like `.redistribute()`, `.to_local()`, and the `DTensor()` constructor, that we want to put directly into the graph so that we can compile them (e.g. redistribute() will desugar into collective ops). Doing this requires functionalization to run **underneath** the subclass though. I hacked around this for now, by forcing these functions to run functionalization first if they need to.
- the backward test that I have is... wrong. The backward graph that we trace out looks kind of reasonable, but it gives incorrect gradients on one of the two inputs. This needs further debugging (presumably we should be able to stare at the graph and identify which part of it is wrong?).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105236

### CODE REVIEW

**Comment by bdhirsh** [comment]
Ready for review! @wanchaol

File: test/distributed/_tensor/test_dtensor.py [hunk 1]
```diff
@@ -704,5 +707,265 @@ def test_split_tensor(self) -> None:
...
+    return [view_8, view_6]""",  # noqa: B950
+        )
+
+        # TODO: unfortunately we are getting incorrect gradients :(
+        # Ideally, we should be able to stare at the backward graph above and figure out what we're computing incorrectly.
+        # y_ref.grad = tensor([0., 1., 2., 3., 4., 5., 6., 7.])
+        # y.grad     = tensor([4., 4., 4., 4., 6., 6., 6., 6.])
+
+        self.assertEqual(x_ref.grad, x.grad)
+        # self.assertEqual(y_ref.grad, y.grad)
```
Comment (bdhirsh) [inline test/distributed/_tensor/test_dtensor.py:967]: darn, @wanchaol looks like we're computing an incorrect gradient in the backward in this test. I know you looked into a flavor of this problem a month or two ago- what was your fix again? I'll poke around, but if you have any ideas let me know!

Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:967]: It's because of fake_pg, we can't rely on data correctness when using fake_pg, so if we want to test out the gradient correctness, we need to use multithreaded pg probably

File: test/distributed/_tensor/test_dtensor.py [hunk 2]
```diff
@@ -704,5 +707,265 @@ def test_split_tensor(self) -> None:
...
+
+        x = torch.ones(1)
+        ref = fn(x)
+        opt_fn = torch.compile(fn, backend="aot_eager", fullgraph=True)
+        res = opt_fn(x)
+        self.assertEqual(res, ref)
+
+    # This is a test where we do the conversions to-from DTensors *outside* of the compiled region,
+    # and the compiled graph has DTensor inputs and outputs.
+    def test_compile_dtensor_simple(self):
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:821]: can we move these test cases to `test_dtensor_compile.py` so that we can group all pt2 related tests? thanks!

File: torch/distributed/_tensor/dispatch.py [hunk 3]
```diff
@@ -150,10 +152,11 @@ def _operator_dispatch(
             args_schema.append(arg._spec)
             local_args.append(arg._local_tensor)
             if mesh is not None:
-                if mesh != arg.device_mesh:
-                    raise NotImplementedError(
-                        f"{op_call}: DTensor does not support cross-mesh operation yet!"
-                    )
+                with disable_proxy_modes_tracing(), unset_fake_temporarily():
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:155]: I think here we don't need to have these two modes enabled anymore, I changed the hashing behavior so that it does not do tensor compare anymore

File: test/distributed/_tensor/test_dtensor.py [hunk 4]
```diff
@@ -704,5 +707,265 @@ def test_split_tensor(self) -> None:
...
+
+        # The bw graph here is pretty messy (there are unnecessary cat and split calls?)
+        # But the important ting to note is:
+        # (1) we have a c10d_functional op in the graph, thanks to DTensor
+        # (2) we have a wait() call in the graph that inductor can reorder
+        # (3) We have a mm() in the graph (from the user code)
+        self.assertExpectedInline(
+            bw_graph_cell[0].code.strip(),
+            """\
+def forward(self, t_1, clone, tangents_1):
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:941]: do we need to assert the graph here? a bit worried that the graph assertion might be hard to navigate when the pt2 stack have new changes breaks the assertion (i.e. adding new op decomp), I am wondering if we should just do `fullgraph=True` and assert the output?

**Review by wanchaol — APPROVED** [review]
a few suggestions about mode unsetting and tests, see inline comments. This looks great in general, thanks for the enablements and exciting to see we can use torch.compile + DTensor!

File: test/distributed/_tensor/test_dtensor.py (hunk 2 above)
Comment (bdhirsh) [reply test/distributed/_tensor/test_dtensor.py:821]: will do!

File: test/distributed/_tensor/test_dtensor.py (hunk 4 above)
Comment (bdhirsh) [reply test/distributed/_tensor/test_dtensor.py:941]: hmm, the nice thing about the graph assert IMO is that someone can at-a-glance look at the tests and understand what the workflow of compiling DTensors is ("this is what matmul + redistribute in the backward desugars into").

If we update a decomp somewhere, updating the tests is pretty easy - running `EXPECTTEST_ACCEPT=1 python test/distributed/_tensor/test_dtensor.py` will automatically update the file.

But if you'd still prefer to keep these out of the test file, lmk and I'm happy to r3emove.

File: test/distributed/_tensor/test_dtensor.py (hunk 1 above)
Comment (bdhirsh) [reply test/distributed/_tensor/test_dtensor.py:967]: wow haha- I knew this and forgot it.

File: torch/distributed/_tensor/dispatch.py (hunk 3 above)
Comment (bdhirsh) [reply torch/distributed/_tensor/dispatch.py:155]: nice, will take out

File: test/distributed/_tensor/test_dtensor.py (hunk 4 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:941]: Oh I see, that's good then, let's keep it! Might worth adding this `EXPECTTEST_ACCEPT=1 python test/distributed/_tensor/test_dtensor.py` command in the test comment so ppl know how to fix those when they saw failures :)

File: torch/distributed/_tensor/sharding_prop.py [hunk 5]
```diff
@@ -124,10 +125,17 @@ def _wrap_output_spec_tensor_meta(
                         spec.tensor_meta = output_tensor_meta_i
 
     def propagate(self, op_info: OpInfo) -> None:
-        output_sharding = self.propagate_op_sharding(op_info.schema)
+        # We cannot use an lru cache if we know that inputs will have dynamic shapes,
+        # because SymInts are not hashable.
+        # This is generally ok because this only happens during tracing in torch.compile,
+        # and tracing does not need to be as fast as eagermode DTensor usages.
+        if _are_we_tracing():
```
Comment (bdhirsh) [inline torch/distributed/_tensor/sharding_prop.py:132]: welp, this isn't enough. It's possible to have the current `op_info.schema` hold onto SymInts, even when `_are_we_tracing()` returns false (because neither a FakeTensorMode or TorchProxyDispatchMode are active).

I'm going to update this check to instead just directly poke at the `op_info.schema` and check if it has any SymInts on it.
